File: packages/unity/Editor/Core/Receiver.cs
using System;
using System.IO;
using System.Security.Cryptography;
using System.Text;
using System.Text.RegularExpressions;

namespace RezonaLab.EngineBridge.Editor
{
    /// <summary>transfer_begin 阶段的校验结果：先限额（4413 类）再一致性（4400 类）再文件名 / 格式（应用层错误，连接保持）。</summary>
    public sealed class BeginRejection
    {
        public enum RejectionKind { Limit, BadFrame, AppError }

        public RejectionKind Kind;
        public string Code;
        public string Message;

        public static BeginRejection Limit(string code, string message) => new BeginRejection { Kind = RejectionKind.Limit, Code = code, Message = message };
        public static BeginRejection BadFrame(string message) => new BeginRejection { Kind = RejectionKind.BadFrame, Message = message };
        public static BeginRejection AppError(string code, string message) => new BeginRejection { Kind = RejectionKind.AppError, Code = code, Message = message };
    }

    public sealed class FrameOrderException : Exception
    {
        public FrameOrderException(string message) : base(message) { }
    }

    public sealed class LimitException : Exception
    {
        public LimitException(string message) : base(message) { }
    }

    /// <summary>
    /// 单次传输的接收器：分块追加写入 &lt;tmpDir&gt;/&lt;transferId&gt;.part，边写边算 sha256，内存占用与文件大小无关；
    /// 全部到齐并校验后原子搬进工程目录。任何失败路径都删 .part。
    /// </summary>
    public sealed class TransferReceiver : IDisposable
    {
        // 文件名里不允许出现路径分隔符、NUL 与其它控制字符、`..`
        private static readonly Regex FileNameForbidden = new Regex(@"[\\/\x00-\x1f\x7f]|\.\.", RegexOptions.Compiled);
        private static readonly Regex UnsafeIdChars = new Regex("[^A-Za-z0-9._-]", RegexOptions.Compiled);

        public string TransferId { get; }
        public string FileName { get; }
        public long ByteSize { get; }
        public long ChunkCount { get; }
        public string PartPath { get; }

        private readonly string _expectedSha;
        private readonly long _chunkBytes;
        private IncrementalHash _hash = IncrementalHash.CreateHash(HashAlgorithmName.SHA256);
        private FileStream _stream;
        private long _received;
        private long _nextIndex;
        private bool _finalized;

        public TransferReceiver(TransferBegin begin, string tmpDir)
        {
            TransferId = begin.TransferId;
            FileName = begin.FileName;
            ByteSize = begin.ByteSize;
            ChunkCount = begin.ChunkCount;
            _chunkBytes = begin.ChunkBytes;
            _expectedSha = begin.Sha256.ToLowerInvariant();
            // transferId 已由 Schema 限制长度；这里再剥掉一切非安全字符，临时文件名不信任任何外部输入
            PartPath = Path.Combine(tmpDir, UnsafeIdChars.Replace(begin.TransferId, "_") + ".part");
        }

        public long BytesReceived => _received;
        public long ExpectedIndex => _nextIndex;
        public int Percent => ByteSize == 0 ? 100 : (int)Math.Floor(_received * 100.0 / ByteSize);

        public static BeginRejection ValidateBegin(TransferBegin msg, Limits limits, string[] formats)
        {
            if (msg.ByteSize > limits.MaxFileBytes)
                return BeginRejection.Limit("TOO_LARGE", "byteSize " + msg.ByteSize + " exceeds maxFileBytes " + limits.MaxFileBytes);
            if (msg.ChunkCount > limits.MaxChunks)
                return BeginRejection.Limit("TOO_MANY_CHUNKS", "chunkCount " + msg.ChunkCount + " exceeds maxChunks " + limits.MaxChunks);
            if (msg.ChunkBytes != limits.ChunkBytes) return BeginRejection.BadFrame("chunkBytes must equal " + limits.ChunkBytes);
            var expectedCount = (msg.ByteSize + msg.ChunkBytes - 1) / msg.ChunkBytes;
            if (msg.ChunkCount != expectedCount) return BeginRejection.BadFrame("chunkCount inconsistent with byteSize");
            if (FileNameForbidden.IsMatch(msg.FileName) || msg.FileName.Trim() != msg.FileName || msg.FileName.StartsWith(".", StringComparison.Ordinal))
                return BeginRejection.BadFrame("fileName contains path separators or control characters");
            var ext = ExtensionOf(msg.FileName);
            if (ext.Length == 0 || Array.IndexOf(formats, ext) < 0)
                return BeginRejection.AppError("UNSUPPORTED_FORMAT", "format ." + (ext.Length == 0 ? "?" : ext) + " not accepted");
            if (msg.Kind == "sprite" && ext != "zip") return BeginRejection.AppError("UNSUPPORTED_FORMAT", "sprite must be a zip");
            return null;
        }

        /// <summary>小写扩展名（不带点）；与 Node 的 extname 一致：以点开头的名字不算有扩展名。</summary>
        public static string ExtensionOf(string fileName)
        {
            var dot = fileName.LastIndexOf('.');
            if (dot <= 0 || dot == fileName.Length - 1) return "";
            return fileName.Substring(dot + 1).ToLowerInvariant();
        }

        public void Open()
        {
            Directory.CreateDirectory(Path.GetDirectoryName(PartPath));
            _stream = new FileStream(PartPath, FileMode.Create, FileAccess.Write, FileShare.None);
        }

        /// <summary>严格顺序：index 必须等于期待的下一块；块大小必须符合计划。违反即抛 FrameOrderException（4400）/ LimitException（4413）。</summary>
        public void WriteChunk(long index, byte[] data, int offset, int length)
        {
            if (_stream == null) throw new InvalidOperationException("receiver not open");
            if (index != _nextIndex) throw new FrameOrderException("expected chunk " + _nextIndex + ", got " + index);
            if (index >= ChunkCount) throw new FrameOrderException("chunk " + index + " beyond chunkCount " + ChunkCount);
            var isLast = index == ChunkCount - 1;
            var expectedLen = isLast ? ByteSize - _chunkBytes * index : _chunkBytes;
            if (length != expectedLen) throw new FrameOrderException("chunk " + index + " has " + length + " bytes, expected " + expectedLen);
            if (_received + length > ByteSize) throw new LimitException("received bytes exceed byteSize");
            _stream.Write(data, offset, length);
            _hash.AppendData(data, offset, length);
            _received += length;
            _nextIndex += 1;
        }

        /// <summary>关闭写流并校验字节数与 sha256。不通过时删 .part 并返回失败原因（null 表示通过）。</summary>
        public string Finish()
        {
            if (_finalized) return "already finalized";
            _finalized = true;
            CloseStream();
            if (_received != ByteSize || _nextIndex != ChunkCount)
            {
                Abort();
                return "byte count mismatch";
            }
            var digest = ToHex(_hash.GetHashAndReset());
            if (digest != _expectedSha)
            {
                Abort();
                return "sha256 mismatch";
            }
            return null;
        }

        /// <summary>把校验过的 .part 搬到 &lt;destDir&gt;/&lt;fileName&gt;；同名存在则加 -2、-3 后缀。File.Move 跨卷时自动退化为 copy + delete。</summary>
        public string MoveTo(string destDir)
        {
            Directory.CreateDirectory(destDir);
            var target = UniquePath(destDir, FileName);
            File.Move(PartPath, target);
            return target;
        }

        /// <summary>删除临时文件（幂等）。连接断开 / 任何失败都要调。</summary>
        public void Abort()
        {
            CloseStream();
            try { if (File.Exists(PartPath)) File.Delete(PartPath); }
            catch (IOException) { /* 尽力而为 */ }
            catch (UnauthorizedAccessException) { /* 尽力而为 */ }
        }

        public void Dispose()
        {
            CloseStream();
            _hash?.Dispose();
            _hash = null;
        }

        /// <summary>&lt;dir&gt;/&lt;name&gt; 不存在则原样返回；否则依次试 &lt;stem&gt;-2&lt;ext&gt;、&lt;stem&gt;-3&lt;ext&gt;…（目录名也适用）。</summary>
        public static string UniquePath(string dir, string name)
        {
            var ext = Path.GetExtension(name);
            var stem = name.Substring(0, name.Length - ext.Length);
            for (var n = 1; n < 10_000; n++)
            {
                var candidate = Path.Combine(dir, n == 1 ? name : stem + "-" + n + ext);
                if (!File.Exists(candidate) && !Directory.Exists(candidate)) return candidate;
            }
            throw new BridgeException("INTERNAL", "cannot find a free name for " + name);
        }

        private void CloseStream()
        {
            if (_stream == null) return;
            var s = _stream;
            _stream = null;
            s.Dispose();
        }

        private static string ToHex(byte[] bytes)
        {
            var sb = new StringBuilder(bytes.Length * 2);
            foreach (var b in bytes) sb.Append(b.ToString("x2"));
            return sb.ToString();
        }
    }
}

File: packages/core-ts/src/receiver.ts
import { createHash, type Hash } from 'node:crypto';
import { promises as fs } from 'node:fs';
import { basename, extname, join } from 'node:path';
import { BridgeError, type Limits, type TransferBeginMessage } from './types.js';

// 文件名里不允许出现路径分隔符、NUL 与其它控制字符、`..`
// eslint-disable-next-line no-control-regex -- 就是要拒绝控制字符
const FILENAME_FORBIDDEN = /[\\/\x00-\x1f\x7f]|\.\./;

/** transfer_begin 阶段的校验结果：先限额（4413 类）再一致性（4400 类）再文件名 / 格式（应用层错误，连接保持）。 */
export type BeginRejection =
  | { kind: 'limit'; code: 'TOO_LARGE' | 'TOO_MANY_CHUNKS'; message: string }
  | { kind: 'bad_frame'; message: string }
  | { kind: 'app_error'; code: 'UNSUPPORTED_FORMAT'; message: string };

export function validateBegin(msg: TransferBeginMessage, limits: Limits, formats: readonly string[]): BeginRejection | null {
  if (msg.byteSize > limits.maxFileBytes) {
    return { kind: 'limit', code: 'TOO_LARGE', message: `byteSize ${msg.byteSize} exceeds maxFileBytes ${limits.maxFileBytes}` };
  }
  if (msg.chunkCount > limits.maxChunks) {
    return { kind: 'limit', code: 'TOO_MANY_CHUNKS', message: `chunkCount ${msg.chunkCount} exceeds maxChunks ${limits.maxChunks}` };
  }
  if (msg.chunkBytes !== limits.chunkBytes) return { kind: 'bad_frame', message: `chunkBytes must equal ${limits.chunkBytes}` };
  if (msg.chunkCount !== Math.ceil(msg.byteSize / msg.chunkBytes)) return { kind: 'bad_frame', message: 'chunkCount inconsistent with byteSize' };
  if (FILENAME_FORBIDDEN.test(msg.fileName) || msg.fileName.trim() !== msg.fileName || msg.fileName.startsWith('.')) {
    return { kind: 'bad_frame', message: 'fileName contains path separators or control characters' };
  }
  const ext = extname(msg.fileName).slice(1).toLowerCase();
  if (!ext || !formats.includes(ext)) return { kind: 'app_error', code: 'UNSUPPORTED_FORMAT', message: `format .${ext || '?'} not accepted` };
  if (msg.kind === 'sprite' && ext !== 'zip') return { kind: 'app_error', code: 'UNSUPPORTED_FORMAT', message: 'sprite must be a zip' };
  return null;
}

export class FrameOrderError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'FrameOrderError';
  }
}
export class LimitError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'LimitError';
  }
}

/**
 * 单次传输的接收器：分块追加写入 `<tmpDir>/<transferId>.part`，边写边算 sha256，
 * 内存占用与文件大小无关；全部到齐并校验后原子搬进工程目录。任何失败路径都删 .part。
 */
export class TransferReceiver {
  readonly transferId: string;
  readonly fileName: string;
  readonly byteSize: number;
  readonly chunkCount: number;
  readonly partPath: string;
  private readonly expectedSha: string;
  private readonly chunkBytes: number;
  private readonly hash: Hash = createHash('sha256');
  private handle: fs.FileHandle | null = null;
  private received = 0;
  private nextIndex = 0;
  private finalized = false;

  constructor(begin: TransferBeginMessage, tmpDir: string) {
    this.transferId = begin.transferId;
    this.fileName = begin.fileName;
    this.byteSize = begin.byteSize;
    this.chunkCount = begin.chunkCount;
    this.chunkBytes = begin.chunkBytes;
    this.expectedSha = begin.sha256.toLowerCase();
    // transferId 已由 Schema 限制长度；这里再剥掉一切非安全字符，临时文件名不信任任何外部输入
    this.partPath = join(tmpDir, `${begin.transferId.replace(/[^A-Za-z0-9._-]/g, '_')}.part`);
  }

  get bytesReceived(): number {
    return this.received;
  }
  get expectedIndex(): number {
    return this.nextIndex;
  }
  get percent(): number {
    return this.byteSize === 0 ? 100 : Math.floor((this.received / this.byteSize) * 100);
  }

  async open(): Promise<void> {
    await fs.mkdir(join(this.partPath, '..'), { recursive: true });
    this.handle = await fs.open(this.partPath, 'w');
  }

  /** 严格顺序：index 必须等于期待的下一块；块大小必须符合计划。违反即抛 FrameOrderError（调用方关 4400）/ LimitError（4413）。 */
  async writeChunk(index: number, data: Buffer): Promise<void> {
    if (!this.handle) throw new Error('receiver not open');
    if (index !== this.nextIndex) throw new FrameOrderError(`expected chunk ${this.nextIndex}, got ${index}`);
    if (index >= this.chunkCount) throw new FrameOrderError(`chunk ${index} beyond chunkCount ${this.chunkCount}`);
    const isLast = index === this.chunkCount - 1;
    const expectedLen = isLast ? this.byteSize - this.chunkBytes * index : this.chunkBytes;
    if (data.length !== expectedLen) throw new FrameOrderError(`chunk ${index} has ${data.length} bytes, expected ${expectedLen}`);
    if (this.received + data.length > this.byteSize) throw new LimitError('received bytes exceed byteSize');
    await this.handle.write(data, 0, data.length);
    this.hash.update(data);
    this.received += data.length;
    this.nextIndex += 1;
  }

  /** 关闭写流并校验字节数与 sha256。不通过时删 .part 并返回失败原因。 */
  async finish(): Promise<{ ok: true } | { ok: false; reason: string }> {
    if (this.finalized) return { ok: false, reason: 'already finalized' };
    this.finalized = true;
    await this.closeHandle();
    if (this.received !== this.byteSize || this.nextIndex !== this.chunkCount) {
      await this.abort();
      return { ok: false, reason: 'byte count mismatch' };
    }
    if (this.hash.digest('hex') !== this.expectedSha) {
      await this.abort();
      return { ok: false, reason: 'sha256 mismatch' };
    }
    return { ok: true };
  }

  /**
   * 把校验过的 .part 搬到 `<destDir>/<fileName>`；同名存在则加 `-2`、`-3` 后缀。
   * 临时目录与工程目录可能不同卷（EXDEV），此时退化为 copy + unlink。返回最终绝对路径。
   */
  async moveTo(destDir: string): Promise<string> {
    await fs.mkdir(destDir, { recursive: true });
    const target = await uniquePath(destDir, this.fileName);
    try {
      await fs.rename(this.partPath, target);
    } catch (err) {
      if ((err as NodeJS.ErrnoException).code !== 'EXDEV') throw err;
      await fs.copyFile(this.partPath, target);
      await fs.unlink(this.partPath);
    }
    return target;
  }

  /** 删除临时文件（幂等）。连接断开 / 任何失败都要调。 */
  async abort(): Promise<void> {
    await this.closeHandle();
    await fs.rm(this.partPath, { force: true });
  }

  private async closeHandle(): Promise<void> {
    if (this.handle) {
      const h = this.handle;
      this.handle = null;
      await h.close();
    }
  }
}

/** `<dir>/<name>` 不存在则原样返回；否则依次试 `<stem>-2<ext>`、`<stem>-3<ext>`…（目录名也适用）。 */
export async function uniquePath(dir: string, name: string): Promise<string> {
  const ext = extname(name);
  const stem = basename(name, ext);
  for (let n = 1; n < 10_000; n++) {
    const candidate = join(dir, n === 1 ? name : `${stem}-${n}${ext}`);
    try {
      await fs.access(candidate);
    } catch {
      return candidate;
    }
  }
  throw new BridgeError('INTERNAL', `cannot find a free name for ${name}`);
}
